[flutter_tools] Removes the need of a no-op plugin implementations

franciscojma86 opened the pull request:
## Description

As mentioned in https://github.com/flutter/flutter/issues/46618, macOS plugins require an iOS no-op implementation and vice versa. This is due to some limitations in the generated `.flutter-plugins` file.

Updates  `.flutter-plugin-dependencies` to include the plugins list and version information. Supports all platforms. Example:
```
{
  "info": "This is a generated file; do not edit or check into version control.",
  "plugins": {
    "ios": [
      {
        "name": "e2e",
        "path": "/path/to/plugin",
      },
      {
        "name": "shared_preferences",
        "path": "/path/to/plugin",
      }
    ],
    "macos": [
      {
        "name": "shared_preferences_macos",
        "path": "/path/to/plugin",
      }
    ],
    "web": [
      {
        "name": "shared_preferences_web",
        "path": "/path/to/plugin",
      }
    ]
  }
}
```

This PR doesn't include changes to the Podfile and gradle templates, since there are still some issues to consider regarding the migration strategy and CI changes.

## Related Issues

https://github.com/flutter/flutter/issues/46618

## Tests

I added the following tests:

Updated tests in `plugins-test.dart` and `plugin-dependencies-test.dart`. Still supports tests for current logic. 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

franciscojma86 committed:
Reland "fix duration event of timeline summary (#47742)"

This reverts commit 50d421224d6b4e48589525be5d9eb036fa20a5c8.

franciscojma86 committed:
Merge branch 'master' of github.com:flutter/flutter

franciscojma86 committed:
Merge branch 'master' of github.com:flutter/flutter

franciscojma86 committed:
Write in json

franciscojma86 committed:
Add all platforms

franciscojma86 committed:
Remove html

franciscojma86 committed:
Podfile changes and add path

franciscojma86 committed:
Update podfiles

franciscojma86 committed:
Fix ios

franciscojma86 committed:
Refactor and add couple of tests

franciscojma86 committed:
Almost all tests pass

franciscojma86 committed:
All refresh tests work

franciscojma86 committed:
Android builds

franciscojma86 committed:
Remove old methods

franciscojma86 committed:
Whitespace

franciscojma86 committed:
Rename api

franciscojma86 committed:
Revert changes in podfile and gradle

franciscojma86 committed:
Documentation

franciscojma86 committed:
Documentation

franciscojma86 committed:
Documentation

franciscojma86 commented:
/cc @amirh

franciscojma86 committed:
Documentation

jonahwilliams commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:async';
 
+import 'package:flutter_tools/src/platform_plugins.dart';
```
This should be a relative import

jonahwilliams commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,38 @@ class FlutterProject {
   }
 }
 
+abstract class FlutterProjectPlatform {
```
Please add a doc comment to this class

jonahwilliams commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,38 @@ class FlutterProject {
 
+abstract class FlutterProjectPlatform {
+
+  // Plugin's platform config key, i.e., "macos", "ios".
```
These `//` should be `///` unless they are only implementation notes

jonahwilliams commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,38 @@ class FlutterProject {
+  bool existsSync();
+
+  // Creates a list with the project plugins supported for the current platform.
+  List<Map<String,dynamic>> pluginsList(List<Plugin>plugins) {
```
```suggestion
  List<Map<String, dynamic>> pluginsList(List<Plugin>plugins) {
```

jonahwilliams commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,38 @@ class FlutterProject {
+    final Iterable<Plugin> platformPlugins = plugins.where((Plugin p) {
+      return p.platforms.containsKey(pluginConfigKey);
+    });
+      final Set<String> pluginNames = <String>{};
```
```suggestion
    final Set<String> pluginNames = <String>{};
```

jonahwilliams replied:
.. I think

franciscojma86 committed:
Documentation from review

jonahwilliams commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
+/// If there aren't any plugins, then the files aren't written to disk. The resulting
+/// file looks something like this:
+/// {
+///     "_info": "// This is a generated file; do not edit or check into version control.",
```
I would also recommend embedding some sort of version string in the JSON, just in case

jonahwilliams replied:
Also possibly the date it was generated on/ the version of the tool used to generate it

jonahwilliams replied:
See .dart_tool/package_config.json:

```
  "generated": "2020-01-13T22:42:48.743606Z",
  "generator": "pub",
  "generatorVersion": "2.8.0-dev.0.0.flutter-395daaa3ec"
```

franciscojma86 replied:
Done

blasten commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
+  }
+}
+
+/// Writes the .flutter-plugins.json file based on the list of plugins.
```
`.flutter-plugins-dependencies` since new projects on stable have this file in the .gitignore already.

franciscojma86 replied:
I changed it to `.flutter-plugins-dependencies`, and won't break since I still add the `dependencyGraph` to support current projects. Once it's fully deprecated we can just remove it from the json

franciscojma86 committed:
Use dependencies file instea

franciscojma86 committed:
Fixed some tests

stuartmorgan commented:
Was there discussion somewhere about using a single file that I missed while out? The plan I had proposed was to make a file per platform, in the platform directories, which seemed like it would be simpler for each platform to parse, as well as sticking to the current structure of platform-specific information generally being within platform directories. I'm curious what the pro/con evaluation of that vs. the approach here was.

jonahwilliams commented:
> Was there discussion somewhere about using a single file that I missed while out?

I think we talked briefly about this before break but I've lost context on the discussion.

blasten commented:
I don’t think I was involved in the discussion about deprecating .flutter-plugins.

Unfortunately, in some situations, Android and iOS have project files that depend on the location and format of this file. I’m looking into addressing this specific problem (managing project files) in a separate effort.

Maybe the rest of the platform can have their own file while iOS/Android continue to use .flutter-plugins until we migrate the affected projects.

jonahwilliams commented:
Actually, I remember now. I think the context was concerns about how updates to the platform directories (if any) would effect this file. @franciscojma86 correct me if wrong

stuartmorgan commented:
> how updates to the platform directories (if any) would effect this file

This file is generated on demand though; I would expect that for desktop it would be in the `ephemeral/` directories, which means not checked in, so I'm not sure what the problem would be.

franciscojma86 commented:
Yeah it was a brief discussion with @jonahwilliams , but I should've opened an issue and discussed there. Looked simple enough at the beginning (doesn't it all? 😄 ).

There were no strong reasons to go either way. The one advantage of going this way, is to stop using the key-value file and turn it into a JSON which is more flexible and we can version. Specially since `.flutter-plugins-dependencies` is already a JSON file. I actually did a per platform implementation here https://github.com/flutter/flutter/pull/47258. It turns out it breaks [other tests](https://cirrus-ci.com/task/6653029600460800?command=main#L115) since it creates the ephemeral folder prematurely. The current implementation has been a bit smoother. 

The only advantage I see of doing a per platform `.flutter-plugins` file is that we wouldn't have to add json parsing logic on the Podfile templates, but I actually already did that in another branch, and it's just a couple extra lines of codes. 

@stuartmorgan do you have any concerns with this method? The other PR is not far behind so I can continue that one if people think is a better options. Personally, I don't see any clear advantages.

stuartmorgan commented:
I have a couple of potential concerns, although nothing definite enough to prevent moving forward:
- The JSON dependency isn't just for Podfiles, it's for every platform that needs to do something with those files. E.g., if I want to use it on Linux, then I need to do JSON parsing in... a tool called from a Makefile I guess? But since I don't have the details of how it'll be used, that's not a concrete concern, since maybe that's no harder than whatever I would need to do with the key-value version. (Also, `.flutter-plugins-dependencies` already being JSON may make that moot, since I may need that on other platforms as well.)
- In a future where platform build details are in separate packages, having them all writing into a single file seems potentially awkward, but that's presumably something we can address by just having one of the APIs be to provide the JSON to put in the platform-specific section.

And I can see the argument for this being potentially more future-proof, since we can add new sections to a JSON file without breaking existing projects. Let's proceed with this, and if it's problematic for Linux or Windows later I can figure out how to adapt.

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
+    }
+  }
+
+  return oldPluginFileContent != _readFileContent(pluginsFile);
```
Since there's encoding/decoding going on now, it would be nice to make this a little bit more efficient. You don't need to `json.encode` above if `pluginsMap.isEmpty`, and you don't need to do the file readback since you already have `pluginFileContent`. There's already branching for empty vs. non-empty, so being more efficient here would add minimal complexity. Something like:

```
  final File pluginsFile = project.flutterPluginsJsonFile;
  final String oldPluginFileContent = _readFileContent(pluginsFile);
  final String pluginFileContent = pluginsMap.isNotEmpty ? json.encode(result) : null;
  if (pluginFileContent != null) {
    pluginsFile.writeAsStringSync(pluginFileContent, flush: true);
  } else {
    if (pluginsFile.existsSync()) {
      pluginsFile.deleteSync();
    }
  }

  return pluginFileContent != oldPluginFileContent;
```

franciscojma86 replied:
I moved the check for empty plugins to the top. Feels a bit cleaner to me. WDYT?

stuartmorgan commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,40 @@ class FlutterProject {
+  /// Plugin's platform config key, i.e., "macos", "ios".
+  String get pluginConfigKey;
+
+  /// Whether a the platform exists in the project.
```
typo: "a the"

stuartmorgan commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,40 @@ class FlutterProject {
+  /// Whether a the platform exists in the project.
+  bool existsSync();
+
+  /// Creates a list with the project plugins supported for the current platform.
```
Maybe "Filters [plugins] to those supported by this platform"?

stuartmorgan commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +255,40 @@ class FlutterProject {
+  bool existsSync();
+
+  /// Creates a list with the project plugins supported for the current platform.
+  List<Map<String, dynamic>> pluginsList(List<Plugin>plugins) {
```
`filterPlugins`? The current name suggests to me that it's returning a list of plugins that are a property of the project, rather than being driven by the input list. (It feels a bit odd for this to be part of the project, rather than a stand-alone utility or something on a plugins class, but I don't feel super strongly about that.)

franciscojma86 replied:
I agree it feels better as a helper function. Done

stuartmorgan commented:
> Maybe the rest of the platform can have their own file while iOS/Android continue to use .flutter-plugins until we migrate the affected projects.

Isn't that why the legacy file is still being written in this patch? We should still be creating the new file for all platforms so there's something to migrate to later.

franciscojma86 committed:
Add to right clock

franciscojma86 committed:
Address some review comments on code improvements

franciscojma86 committed:
Move filterPlugins as a helper

franciscojma86 commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
 }
 
+void _addPluginsToPlatform(List<Plugin> plugins, FlutterProjectPlatform platform, Map<String, dynamic> pluginsMap) {
+  if (!platform.existsSync()) {
```
I got rid of this and added to the map directly. I was trying to avoid writing platforms that have no plugins but as you say, there's no much point to that. If there are no plugins at all, the file won't be written.

franciscojma86 commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
   return plugins;
 }
 
+void _addPluginsToPlatform(List<Plugin> plugins, FlutterProjectPlatform platform, Map<String, dynamic> pluginsMap) {
```
Removed it

franciscojma86 commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,20 +302,105 @@ List<Plugin> findPlugins(FlutterProject project) {
+  const String info = 'This is a generated file; do not edit or check into version control.';
+
+  final Map<String, dynamic> result = <String, dynamic> {};
+  result['_info'] = '// $info';
```
Bad copy paste :)

franciscojma86 commented:
@stuartmorgan I changed the contents to `.flutter-plugins-dependencies` to avoid creating a new file and doing a bunch of CI work. PTAL!

franciscojma86 committed:
Analyzer

franciscojma86 committed:
Merge branch 'master' of github.com:flutter/flutter

franciscojma86 committed:
Rebased and updated escapePath

franciscojma86 committed:
Mock flutter version

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,63 +304,170 @@ List<Plugin> findPlugins(FlutterProject project) {
 
-/// Writes the .flutter-plugins and .flutter-plugins-dependencies files based on the list of plugins.
-/// If there aren't any plugins, then the files aren't written to disk.
+  /// Filters [plugins] to those supported by this platform.
```
s/this platform/[platformKey]/

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,63 +304,170 @@ List<Plugin> findPlugins(FlutterProject project) {
+    final Set<String> pluginNames = <String>{};
+    for (final Plugin plugin in plugins) {
+      pluginNames.add(plugin.name);
+    }
```
Optional:
```final Set<String> pluginNames = plugins.map((plugin) => plugin.name).toSet();```

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,63 +304,170 @@ List<Plugin> findPlugins(FlutterProject project) {
+/// If there aren't any plugins, then the files aren't written to disk. The resulting
+/// file looks something like this:
+/// {
+///   "info": "This is a generated file; do not edit or check into version control.",
```
Isn't the location of this line in the file going to be "random" (not from run to run unless the map implementation changes, but not guaranteed to be first)? I don't have a solution that doesn't break existing apps since we're re-using a file that's already a dictionary--otherwise we could wrap the real content in an array with the comment then the content--but I'm wondering how useful this will end up being without an order dependency. I guess better than nothing...

franciscojma86 replied:
I changed the test to not require an order comparison and clarified in the comment that the order is not guaranteed

stuartmorgan commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,63 +304,170 @@ List<Plugin> findPlugins(FlutterProject project) {
+  return directAppDependencies;
+}
+
+// This method will be DEPRECATED in favor of _writeFlutterPluginsList.
```
Since this isn't an API surface, it won't ever actually be deprecated, will it, just removed? The API is the existence of the file, not the method.

franciscojma86 replied:
Updated

stuartmorgan commented on packages/flutter_tools/lib/src/project.dart:
```diff
@@ -251,6 +252,16 @@ class FlutterProject {
+/// Base class for projects per platform.
+abstract class FlutterProjectPlatform {
+
+  /// Plugin's platform config key, i.e., "macos", "ios".
```
Nit: s/i.e./e.g./

stuartmorgan commented on packages/flutter_tools/test/general.shard/plugins_test.dart:
```diff
@@ -281,26 +330,85 @@ dependencies:
         );
         expect(flutterProject.flutterPluginsDependenciesFile.readAsStringSync(),
```
Do we actually want this test to break if the order of the elements in this file change? I.e., do we consider the order of elements in a JSON dictionary to be an API? That seems extreme; I know in some cases Ian has advocated for that, but for a file that's only intended for use as JSON I'm not sure we actually want that.

If we don't, this test should parse and then do a deep structure equality test, instead of comparing strings.

franciscojma86 replied:
Updated test as discussed

stuartmorgan-g reviewed (approved):
Just small nits; LGTM

franciscojma86 committed:
Update dependency test and remove dependencies from each plugin

franciscojma86 committed:
Update test without order

franciscojma86 commented on dev/devicelab/bin/tasks/plugin_dependencies_test.dart:
```diff
@@ -74,10 +75,6 @@ Future<void> main() async {
 
-      // https://github.com/flutter/flutter/issues/46898
-      // https://github.com/flutter/flutter/issues/39657
-      File(path.join(pluginCDirectory.path, 'android', 'build.gradle')).deleteSync();
```
Removed this line since the top issue is closed and the second one I assume would be fixed by this PR.

franciscojma86 committed:
space

franciscojma86 committed:
Update test

franciscojma86 commented:
@blasten @jonahwilliams I updated it to include the plugins list in `.flutter-plugins-dependencies` rather than replacing it. Updated a device lab test as well. PTAL!

blasten commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -302,63 +304,156 @@ List<Plugin> findPlugins(FlutterProject project) {
+///     "windows": [],
+///     "web": []
+///   },
+///   "dependencyGraph": [
```
Is the plan to split "dependencyGraph" per platform in a separate PR?

franciscojma86 replied:
I don't think it makes much sense to split it anymore. It doesn't really add any value. I started that way, but then we'd have change the way the dependencies are handled in gradle, which at that point felt unnecessary. What do you think?

blasten replied:
Gradle is working around the fact that `dependencyGraph` isn't split per platform.  I think it would be ideal if Gradle didn't have to workaround that and instead just read this file. This could be done in a separate PR though if that makes sense.

franciscojma86 replied:
Makes sense, I did it here :)

blasten commented on packages/flutter_tools/lib/src/plugins.dart:
```diff
@@ -782,8 +877,13 @@ Future<void> _writeWebPluginRegistrant(FlutterProject project, List<Plugin> plug
 void refreshPluginsList(FlutterProject project, {bool checkProjects = false}) {
   final List<Plugin> plugins = findPlugins(project);
+
+  // TODO(franciscojma): Remove once migration is complete.
```
nit: file a bug for this TODO if there isn't one already.

franciscojma86 committed:
Add issue

blasten reviewed (approved):
LGTM! + comment

franciscojma86 committed:
Add dependencies to plugins

franciscojma86 committed:
Fix plugin filtering
